### ISSUE TITLE
[Bug] ttnn unary ops accept a preallocated output_tensor of the wrong shape: the shape guard compares the shape against itself

### ISSUE THREAD

**Issue by badnikhil** [tenstorrent/tt-metal#55800]
### Describe the bug

The `TT_FATAL` that is supposed to reject a preallocated `output_tensor` whose shape does not match the input compares the preallocated shape **against itself**, so it can never fire. Every `ttnn` unary op therefore accepts a mismatched `output_tensor` silently: an oversized one makes the kernel read tiles from past the end of the input buffer, an undersized one silently returns a truncated result.

[`unary_device_operation.cpp#L101-L107`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp#L101-L107):

```cpp
if (output_tensor.has_value()) {
    const auto computed_output_shape = compute_output_specs(args, tensor_args).logical_shape();
    const auto preallocated_output_shape = output_tensor->logical_shape();
    TT_FATAL(
        preallocated_output_shape == computed_output_shape,
        "Unary: Preallocated output shape must match computed shape. ...");
```

and twenty lines below, [`compute_output_specs`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp#L118-L122) begins:

```cpp
tt::tt_metal::TensorSpec UnaryDeviceOperation::compute_output_specs(
    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
    if (tensor_args.output_tensor.has_value()) {
        return tensor_args.output_tensor->tensor_spec();
    }
```

The guard only runs when `output_tensor.has_value()`, and in exactly that case `compute_output_specs` returns the preallocated tensor's own spec — so the assertion is `output_tensor->logical_shape() == output_tensor->logical_shape()`. The layout check on the next line is written correctly (it compares against `input_tensor.layout()`), which is why a `ROW_MAJOR` output *is* rejected; only the shape arm is dead.

What gets past it is not harmless. The work split is sized from the **output**, [`unary_program_factory.cpp#L206-L207`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_program_factory.cpp#L206-L207):

```cpp
const uint32_t out_num_tiles =
    rm_interleaved ? (k.total_rows + k.rows_per_tile - 1) / k.rows_per_tile : output.physical_volume() / tile_hw;
```

so a 4-tile output makes the reader pull 4 pages out of a 1-page input buffer.

### Repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

def t(shape, val):
    return ttnn.from_torch(torch.full(shape, val, dtype=torch.float32),
                           dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

x = t([1, 1, 32, 32], 2.0)          # 1 page
out = t([1, 1, 64, 64], -99.0)      # 4 pages, prefilled with a sentinel

r = ttnn.exp(x, output_tensor=out)  # expected: TT_FATAL. actual: runs.
print(r.shape)                                                  # Shape([1, 1, 64, 64])
print(sorted(set(ttnn.to_torch(out).flatten().tolist()))[:4])   # [1.0, 7.40625]

# 7.40625 = bf16 exp(2), in 1024 of the 4096 elements -- the one tile that exists.
# 1.0 = exp(0), in the other 3072 -- read from past the end of the 1-page input buffer.
# No element still holds the -99 sentinel, so all four tiles were written.

ttnn.add(x, x, output_tensor=out)   # control: binary_ng raises TT_FATAL here
```

### Observed vs expected

ttsim Blackhole v1.10.6, `bfloat16`, `TILE_LAYOUT`, input filled with `2.0`, `output_tensor` prefilled with the sentinel `-99.0`. This is a validation bug, not a numerics one, so the interesting column is what the caller's buffer holds rather than a ULP count.

| # | call | got | want |
|---|---|---|---|
| A | `ttnn.exp(x[1,1,32,32], output_tensor=out[1,1,64,64])` | no error; returns `[1,1,64,64]` backed by `out`; 1024/4096 elements are `7.40625` (`exp(2)`), 3072 are `1.0` (`exp(0)`, i.e. zeros read past the end of the 1-page input buffer), 0 still hold the sentinel | `TT_FATAL` "Unary: Preallocated output shape must match computed shape" |
| B | `ttnn.exp(x[1,1,64,64], output_tensor=out[1,1,32,32])` | no error; returns `[1,1,32,32]`, all 1024 elements `7.40625` — three quarters of the requested result silently dropped | same `TT_FATAL` |
| C | `ttnn.exp(x[1,1,32,64], output_tensor=out[1,1,64,32])` | no error; the returned tensor carries the caller's shape, not the input's | same `TT_FATAL` |
| D | `ttnn.relu(x[1,1,32,32], output_tensor=out[1,1,32,96])` | no error; all 3072 elements read `2.0` — the two out-of-range tiles came back as `2.0` in this run where they came back as `0.0` in run A | same `TT_FATAL` |
| E | `ttnn.add(x[1,1,32,32], x, output_tensor=out[1,1,64,64])` (control) | `TT_FATAL` from `binary_ng_device_operation.cpp` | correct today |

Run D is the point about the memory safety: the contents of the out-of-range tiles depend on unrelated allocations, not on the input, which is the signature of an out-of-bounds read rather than a defined value.

### Affected range

Every `ttnn` unary op that accepts `output_tensor=` — the whole `UnaryDeviceOperation` family (`exp`, `relu`, `sqrt`, `neg`, `abs`, `tanh`, `gelu`, `typecast`, …), every dtype, every layout. It is host-side device-op C++, so no kernel or SFPU code is involved.

The same idiom — validate against `compute_output_specs()`, which early-returns the preallocated tensor's spec — appears in four more device ops and is dead in all of them:

| file | line | verified |
|---|---|---|
| `ttnn/cpp/ttnn/operations/eltwise/ternary/device/ternary_device_operation.cpp` | 379 (guard), 464 (early return) | yes — `ttnn.where(x[1,1,32,32], x, x, output_tensor=out[1,1,64,64])` runs and returns a `[1,1,64,64]` tensor whose three extra tiles are read out of range |
| `ttnn/cpp/ttnn/operations/experimental/dropout/device/dropout_device_operation.cpp` | 69 (guard), 88 (early return) | by inspection |
| `ttnn/cpp/ttnn/operations/experimental/unary_backward/gelu_backward/device/gelu_backward_device_operation.cpp` | 98 (guard), 111 (early return) | by inspection |
| `ttnn/cpp/ttnn/operations/experimental/ssm/repeat_and_interleave_eltwise_mul/device/repeat_and_interleave_eltwise_mul_device_operation.cpp` | 79 (guard), 96 (early return) | by inspection |

`binary_ng` is the one family that gets it right, and only because it does the comparison *inside* `compute_output_specs`, against the freshly computed broadcast shape, before the early return.

### Model impact

None known, and stated up front: no in-tree model passes a mismatched `output_tensor` on purpose, so nothing in the repo is silently producing wrong numbers today. The value of fixing it is that a caller's shape mistake becomes the error the code already meant to raise instead of an out-of-bounds read whose result depends on what else is resident in DRAM.

### Expected behavior

A preallocated `output_tensor` whose logical shape differs from the shape the op would have produced should be rejected with the `TT_FATAL` that is already written, the same way `binary_ng` rejects it today. Unary ops are elementwise, so the expected shape is the input's logical shape — which is exactly what `compute_output_specs` computes when no output tensor is preallocated.

### Which architectures

All of them. This is host-side device-op validation in `ttnn/cpp`, not a kernel, so Wormhole, Blackhole and Quasar are affected identically.

### How it was verified

ttsim Blackhole v1.10.6 (`TT_METAL_SIMULATOR=libttsim_bh.so`, slow dispatch), `bfloat16`, `TILE_LAYOUT`, with the `output_tensor` prefilled with `-99.0` so that "written" and "left alone" can be told apart, and `buffer_address()` compared to confirm the returned tensor is the caller's buffer. `ttnn.add` and a `ROW_MAJOR` output were used as controls to show the surrounding validation does fire.
